# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [28]:
# прежде чем искать зависимость между количеством детей и возвратом кредита в срок, необходимо обозначить общее количество user с долгом и user без долга
# print(data.groupby('debt').count())
# для проверки строк в датафрейм 21331
# из датафрейм видно, что количество "должников" намного меньше, чем количество людей, не имеющих долгов. 

# проверка зависимости между количеством детей и возвратом кредита в срок

# необходимо построить сводную таблицу количества заёмщиков, количества должников, доли невозврата, группировать необходимо по столбцу children с помощью метода pivot_table
data_pivot1 = data.pivot_table('debt', index = ['children'], aggfunc = ['count', 'sum', 'mean'])
# print(data_pivot1)

data_pivot2 = pd.DataFrame({
  'Количество заемщиков': [14091, 4808, 2052, 330, 41, 9],
  'Количество должников': [1063, 444, 194, 27, 4, 0],
  'Доля невозврата': [0.075438,  0.092346, 0.094542, 0.081818, 0.097561, 0.000000]
})

data_pivot2.index.rename('Количество детей', inplace = True)
print(data_pivot2, '\n') 

#cравнение долей невозврата по двум категориями 
# print('Средняя доля невозврата в категории людей от 0 до 2 детей включительно:', (data_pivot2['Доля невозврата'][0]+data_pivot2['Доля невозврата'][1]+data_pivot2['Доля невозврата'][2])/3)
# print('Средняя доля невозврата в категории людей от 3 до 5 детей включительно:', (data_pivot2['Доля невозврата'][3]+data_pivot2['Доля невозврата'][4]+data_pivot2['Доля невозврата'][5])/3)

with_children = data.loc[data['children'] > 0]['debt'].mean()
print('Доля невозврата в категории для людей, имеющих от 1 и более ребенка:', with_children)
without_children = data.loc[data['children'] <= 0]['debt'].mean()
print('Доля невозврата в категории для людей, не имеющих детей:', without_children)



                  Количество заемщиков  Количество должников  Доля невозврата
Количество детей                                                             
0                                14091                  1063         0.075438
1                                 4808                   444         0.092346
2                                 2052                   194         0.094542
3                                  330                    27         0.081818
4                                   41                     4         0.097561
5                                    9                     0         0.000000 

Доля невозврата в категории для людей, имеющих от 1 и более ребенка: 0.09240331491712707
Доля невозврата в категории для людей, не имеющих детей: 0.07543822297920659


**Вывод:** Зависимость между количеством детей и возвратом кредита в срок есть. Для категории людей, не имеющих детей, доля невозвратов составляет 7.54%, для категории людей, имеющих от 1 и более ребенка доля невозвратов составляет 9.24%
То есть категория людей, которая больше всего несет риски для заказчика: люди, имеющие от 1 и более детей. 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
# проверка зависимости между семейным положением и возвратом кредита в срок
data_pivot = data.pivot_table('debt', index = ['family_status'], aggfunc = ['count', 'sum', 'mean'])
data_pivot = data_pivot.rename(columns={'count': 'Количество заемщиков', 'sum': 'Количество должников', 'mean': 'Доля невозврата'})
data_pivot.index.rename('Семейное положение', inplace = True)
print(data_pivot)

                      Количество заемщиков Количество должников  \
                                      debt                 debt   
Семейное положение                                                
Не женат / не замужем                 2796                  273   
в разводе                             1189                   84   
вдовец / вдова                         951                   63   
гражданский брак                      4134                  385   
женат / замужем                      12261                  927   

                      Доля невозврата  
                                 debt  
Семейное положение                     
Не женат / не замужем        0.097639  
в разводе                    0.070648  
вдовец / вдова               0.066246  
гражданский брак             0.093130  
женат / замужем              0.075606  


**Вывод:** Зависимость между семейным положением и возвратом кредита в срок отследить сложно, так как в данном случае наличие/отсутствие супруга/супруги не даёт гарантию возврата кредита в срок. Категории, которые больше всего несут риска для заказчика: "не женат/ не замужем", доля невозвратов в этой категории составляет 0.0976; "гражданский брак", доля невозвратов в этой категории составляет 0.0931. 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
# на основе ранее созданного столбца total_income_category с помощью функции categorize_income 

#0–30000 — 'E';
#30001–50000 — 'D';
#50001–200000 — 'C';
#200001–1000000 — 'B';
#1000001 и выше — 'A'.

data_pivot = data.pivot_table('debt', index = ['total_income_category'], aggfunc = ['count', 'sum', 'mean'])
data_pivot = data_pivot.rename(columns={'count': 'Количество заемщиков', 'sum': 'Количество должников', 'mean': 'Доля невозврата'})
data_pivot.index.rename('Уровень дохода', inplace = True)
print(data_pivot)

               Количество заемщиков Количество должников Доля невозврата
                               debt                 debt            debt
Уровень дохода                                                          
A                                25                    2        0.080000
B                              5014                  354        0.070602
C                             15921                 1353        0.084982
D                               349                   21        0.060172
E                                22                    2        0.090909


**Вывод:** Категория "Е" и "С" с доходом от 0-30.000 и с доходом 50001-200.000 несет больше всего рисков для заказчика, доля невозвратов в этой категории составляет 0.0909 и 0.0849 соответственно. 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
# группировка по purpose_category и выбор столбца debt

data.groupby('purpose_category')['debt'].sum()

purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

**Вывод:** Больше всего "должников" в категории человек, которые берут кредиты на операции с недвижимостью (780 чел.), далее операции с автомобилем (400 чел.). Предположительно, стоимость "цели", которая предпологает операции с недвижимостью и операции с автомобилем, является наиболее дорогостоящей из всех перечисленных, а значит и влечет за собой более высокие финансовые затраты, следовательно, наибольшая вероятность образования долга прослеживается в этих категориях. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски бывает трех типов: полностью случайные (пропуск не зависит ни от каких-либо других значений), случайные (пропуск зависит от других значений в датафрейме, но не от собственного столбца) и неслучайные (пропуск зависит от значений датафрейма, в том числе и от значений собственного столбца). 
Причинами появления пропусков в исходных данных могут быть: человеческий фактор, потеря данных, ошибка в записи данных. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Если в датафрейме присутствуют выдающие значения, то для заполнения пропусков лучше использовать медиану. Использовав медиану, можно избежать некорректного отображения данных. 

### Шаг 4: общий вывод.

В ходе выполненного проекта проверены гипотезы зависимости различных факторов (количество детей, семейное положение, уровень дохода, цели кредита) на возврат кредита в срок. 
Сделаны следующие выводы: 
1)Зависимость между количеством детей и возвратом кредита в срок есть. Для категории людей, не имеющих детей, доля невозвратов составляет 7.54%, для категории людей, имеющих от 1 и более ребенка доля невозвратов составляет 9.24% То есть категория людей, которая больше всего несет риски для заказчика: люди, имеющие от 1 и более детей.
2)Зависимость между семейным положением и возвратом кредита в срок отследить сложно, так как в данном случае наличие/отсутствие супруга/супруги не даёт гарантию возврата кредита в срок.  Категории, которые больше всего несут риска для заказчика: "не женат/ не замужем", доля невозвратов в этой категории составляет 9.76%. 
3)Категория "Е" и "С" с доходом от 0-30.000 и с доходом 50001-200.000 несет больше всего рисков для заказчика, доля невозвратов в этой категории составляет 9.09% и 8.49% соответственно. 
4)Больше всего "должников" в категории людей, которые берут кредиты на операции с недвижимостью (780 чел.), далее операции с автомобилем (400 чел.).